ip2asn.py

```diff
@@ -0,0 +1,172 @@
+				else:
+					asn_description[as_num] = as_description
+
+				#print(first_ip + ',' + last_ip + ',' + as_num + ',' + country + ',' + as_description)
+				if as_num == '0' and as_description == 'Not routed':
+					pass
+				#elif as_num == '0' and as_description != 'Not routed':
+				#	sys.stderr.write('as == 0, desc != not routed\n')
+				#elif as_num != '0' and as_description == 'Not routed':
+				#	sys.stderr.write('as != 0, desc == not routed\n')
+				else:
+					if first_ip.find(':') > -1:
+						first_addr = IPv6Address(first_ip)
+						last_addr = IPv6Address(last_ip)
+					else:
+						first_addr = IPv4Address(first_ip)
+						last_addr = IPv4Address(last_ip)
+					#except:				# (AddressValueError, ipaddress.AddressValueError):
+
+					sar_cidrs = list(summarize_address_range(first_addr, last_addr))
+					for one_cidr in sar_cidrs:
+						if one_cidr.version == 4:
+							first_octet = '4_' + one_cidr.exploded.split('.')[0]
+						elif one_cidr.version == 6:
+							first_octet = '6_' + one_cidr.exploded[0:2]
+
+						if first_octet not in as_info_struct:
+							as_info_struct[first_octet] = []
+							for cidr_len in range(0, 129):		#Numbers 0 to 128
+								as_info_struct[first_octet].append({})
+
+						one_netmask = one_cidr.prefixlen
+						#print("Prefixlen: " + str(one_netmask))
+						if one_cidr in as_info_struct[first_octet][one_netmask]:
+							if Paranoid and as_info_struct[one_netmask][one_cidr] != as_num:
+								sys.stderr.write("For subnet " + str(one_cidr) + ", " + as_info_struct[one_netmask][one_cidr] + " != " + str(as_num) + "\n")
+						else:
+							as_info_struct[first_octet][one_netmask][one_cidr] = as_num
+	else:
+		sys.stderr.write("ASN Source file " + source_file + " does not exist, unable to lookup ASNs.\n")
+
+	return as_info_struct, asn_country, asn_description
+
+
+
+def ip_asn_lookup(ip_string, as_num_d):
+	"""Find the ASN for the given IP address or None if no match found.  This returns the most specific subnet in case there are multiple matching cidr blocks."""
+	#Approx 0.7 secs/lookup
+
+	if sys.version_info < (3, 0):
+		ip_string = unicode(ip_string)
+
+	if ip_string.find(':') > -1:
+		try:
+			lookup_obj = IPv6Address(ip_string)
+			first_octet = '6_' + lookup_obj.exploded[0:2]
+		except:
+			return None
+		max_index = 128
+	else:
+		try:
+			lookup_obj = IPv4Address(ip_string)
+			first_octet = '4_' + lookup_obj.exploded.split('.')[0]
+		except:
+			return None
+		max_index = 32
+
+
+	if first_octet in as_num_d:
+		for search_netmask in range(max_index, -1, -1):
+			for one_net in as_num_d[first_octet][search_netmask]:
+				if lookup_obj in one_net:
+					return as_num_d[first_octet][search_netmask][one_net]
+	#else:
+	#	return None
+
+	return None
+
+
+
+def formatted_asn_output(orig_ip_string, out_format, as_num_d, as_country_d, as_descriptions_d):
+	"""Take supplied ip string, look up its ASN, and return a formatted output string."""
```

Comment:
Please update the comment to mention that it may return a list of dictionaries or a list of strings.